xtask/src/hooks.rs

```diff
@@ -0,0 +1,161 @@
+use std::fs::File;
+use std::io::{BufRead, BufReader, ErrorKind};
+use std::path::Path;
+use std::process::Command;
+
+use clap::Args;
+use cross::CommandExt;
+
+const CARGO_FLAGS: &[&str] = &["--all-features", "--all-targets", "--workspace"];
+
+#[derive(Args, Debug)]
+pub struct Check {
+    /// Provide verbose diagnostic output.
+    #[clap(short, long)]
+    verbose: bool,
+    /// Run shellcheck on all files, not just staged files.
+    #[clap(short, long)]
+    all: bool,
+}
+
+#[derive(Args, Debug)]
+pub struct Test {
+    /// Provide verbose diagnostic output.
+    #[clap(short, long)]
+    verbose: bool,
+}
+
+fn has_nightly(verbose: bool) -> cross::Result<bool> {
+    Command::new("cargo")
+        .arg("+nightly")
+        .run_and_get_output(verbose)
+        .map(|o| o.status.success())
+        .map_err(Into::into)
+}
+
+fn get_channel_prefer_nightly(
+    verbose: bool,
+    toolchain: Option<&str>,
+) -> cross::Result<Option<&str>> {
+    Ok(match toolchain {
+        Some(t) => Some(t),
+        None => match has_nightly(verbose)? {
+            true => Some("nightly"),
+            false => None,
+        },
+    })
+}
+
+fn cargo(channel: Option<&str>) -> Command {
+    let mut command = Command::new("cargo");
+    if let Some(channel) = channel {
+        command.arg(&format!("+{channel}"));
+    }
+    command
+}
+
+fn cargo_fmt(verbose: bool, channel: Option<&str>) -> cross::Result<()> {
+    cargo(channel)
+        .args(&["fmt", "--", "--check"])
+        .run(verbose, false)
+        .map_err(Into::into)
+}
+
+fn cargo_clippy(verbose: bool, channel: Option<&str>) -> cross::Result<()> {
+    cargo(channel)
+        .arg("clippy")
+        .args(CARGO_FLAGS)
+        .args(&["--", "--deny", "warnings"])
+        .run(verbose, false)
+        .map_err(Into::into)
+}
+
+fn cargo_test(verbose: bool, channel: Option<&str>) -> cross::Result<()> {
+    cargo(channel)
+        .arg("test")
+        .args(CARGO_FLAGS)
+        .run(verbose, false)
+        .map_err(Into::into)
+}
+
+fn splitlines(string: String) -> Vec<String> {
+    string.lines().map(|l| l.to_string()).collect()
+}
+
+fn staged_files(verbose: bool) -> cross::Result<Vec<String>> {
+    Command::new("git")
+        .args(&["diff", "--cached", "--name-only", "--diff-filter=ACM"])
+        .run_and_get_stdout(verbose)
+        .map(splitlines)
+        .map_err(Into::into)
+}
+
+fn all_files(verbose: bool) -> cross::Result<Vec<String>> {
+    Command::new("git")
+        .arg("ls-files")
+        .run_and_get_stdout(verbose)
+        .map(splitlines)
+        .map_err(Into::into)
+}
+
+fn is_shell_script(path: impl AsRef<Path>) -> cross::Result<bool> {
+    if path.as_ref().is_dir() {
+        // is a directory if a git submodule
+        return Ok(false);
+    }
+    let file = File::open(path.as_ref())?;
+    let reader = BufReader::new(file);
+
+    match reader.lines().next() {
+        Some(Ok(line)) => Ok(line.starts_with("#!") && line.contains("sh")),
```

Comment:
A more sophisticated way would be to actually evaluate the shebang line, but this is identical to what we did before, and it's going to work unless we **really** try to mess it up, so I think this is fine.